### ISSUE TITLE
Dask task graphs are huge (non-performant) for large data

### ISSUE THREAD

**Issue by GFleishman** [thewtex/ngff-zarr#48]
I think this will be a tougher issue to solve than typos, math bugs, and formatting; so it will warrant some conversation with @thewtex and @tbirdso.

I have `ngff-zarr` running well on small in-memory sized datasets (e.g. 512x512x512 voxels). When formatted with the nested directory style dimension separator these files open just fine with the napari-ome-zarr plugin (just my first check for compatibility with other tools - I'll use other visualization software later as well). So my next task was to run on real data.

My real image is {'z': 1596, 'y': 8983, 'x': 5451} voxels and I am attempting the following:

```python
ngff_image = to_ngff_image(
    my_big_zarr_array,
    dims=('z', 'y', 'x'),
    scale={a:b for a, b in zip('zyx', my_voxel_spacing)},
    axes_units={a:'micrometer' for a in 'zyx'},
)

multiscales = to_multiscales(
    ngff_image,
    scale_factors=[{'z':4, 'y':8, 'x':8}],
    chunks=(64, 128, 128),
)

to_ngff_zarr(
    './my_data_as_ome_ngff_zarr.zarr',
    'multiscales,
)
```

Importantly, this does actually work. If I wait long enough I get the file I want. However there is a performance issue that IMO will be lethal for the package if not resolved. For this (4, 8, 8) downsample level, the task graph has about 2.5M tasks. The dask distributed scheduler really can't handle graphs of this size well, which it acknowledges with a warning that the task graph itself is about 220MB when compressed, which it knows is huge, and says it will cause long slow downs.

It takes about 45 minutes for the scheduler to finish parsing all those tasks (just about right with the 2.5M tasks and the [documented 0.001 seconds per task](https://distributed.dask.org/en/latest/efficiency.html)), and during this time the workers are allocated but idle. And this has to happen for every scale level. In a real scenario I might have 10 or more scale levels I want to create. So that's a huge amount of time to be paying for cpus that are just waiting for the scheduler to finish parsing the task graph.

I feel that there are three different options for approaching this problem:
(1) small hacks, maybe something will help
(2) optimize the task graph by reworking the series of dask array operations to simplify simplify simplify
(3) abandon dask in as many places as possible

I tried three different things for (1):
(a) ran `dask.array.optimize` on the big graph (only eliminated about 10K tasks and at the expense of running this function which was long)
(b) forced even in-memory size scale levels to run through [the serialization stuff](https://github.com/thewtex/ngff-zarr/blob/92ae33705cd0634e5ec5bdb3a2e233eff6fa2af8/ngff_zarr/to_ngff_zarr.py#L84C9-L84C9) (resulted in 6 task graphs each about 500K tasks, so the total time when breaking the problem up was even longer)
(c) forced every scale to run through this ["Minimize task graph depth" block](https://github.com/thewtex/ngff-zarr/blob/92ae33705cd0634e5ec5bdb3a2e233eff6fa2af8/ngff_zarr/to_ngff_zarr.py#L199) (just resulted in the same graph as before)

So now I want to decide between options (2) and (3). I'd love for there to be a solution with (2). It would be great to stay consistent with the sort of intended dask workflow and also to not have to refactor your package too much. But while building my package [bigstream](https://github.com/GFleishman/bigstream), and running into the same kind of huge task graph problems, I ultimately settled on (3) and everything got really performant really fast. After all, we're just doing a bunch of smoothing and downsampling here. It's not rocket science. Having all of that coordinated by a scheduler is nice, but not if it turns the program into a bureaucracy.

So, sorry for the long message, but I'd love to discuss this with you guys to determine where to go next.

**Comment by thewtex** [comment]
@GFleishman thanks for reporting this and the discussion!

You are right -- dask requires that the task graph it is fed is not too large, too deep, and contains a topology that can parallelize well to work with large data in practices. 

I do have a local WIP branch that improves current situation. After complete, I would love to get your feedback on whether it helps with your issues and brainstorm together on what we can do better.

**Comment by GFleishman** [comment]
@thewtex I'd be happy to help if you want to commit any of that branch. Your understanding of all the dask features is probably better than mine, since with bigstream I basically threw in the towel for a lot of these cases by avoiding the scheduler and having my functions load their own data/overlaps and things like that. But I can at least test in parallel with your development and try to optimize with you.

**Comment by toloudis** [comment]
Has there been any progress on this issue?  I am now trying a similar sequence of functions starting with a dask delayed array that I am stacking together from several images.  The total size is approximately T=500, Z=150, X=2000, Y=2000.  My goal is to generate a zarr that is chunked per XY slice.  In fact, ideally I want to chunk single slices at full res, then 4 slices when xy is scaled down by half, etc.  But I'd settle for one chunk = one slice.

With the current code, I am hitting this error:
```
    multiscales = to_multiscales(
                  ^^^^^^^^^^^^^^^
RuntimeError: Cannot store into in memory Zarr Array using the distributed scheduler.
```
so in my case it's not even getting as far as starting to set up the downsampling task graph.

**Comment by thewtex** [comment]
Hi @toloudis , I am taking steps to first make generation faster, then come back to this.

That said, I think the existing code should be able to process your dataset.

> RuntimeError: Cannot store into in-memory Zarr Array using the distributed scheduler.

Are you using the CLI? Or the API but writing into a zarr memory store? The CLI by default does not load the data into a zarr memory store. It incrementally writes to the disk according to what can be stored in memory. It will run in multiple stages, e.g., if your data is stored as 2000 by 2000 slices, it will first chunk the slice into smaller chunks, loading as many slices as possible without running out of memory, then create 3D chunks from the smaller pieces. The optimization I referenced in this issue does more of these incremental steps.

With the CLI, the flags `--chunks` flag enables specifying the chunk dimension, i.e. to specify XY chunks of a certain size. Also relevant are `--memory-target`, `--local-cluster` flags.

**Comment by toloudis** [comment]
I more or less ported what the cli is doing into my own script because the source data has to be assembled by python code.
I am using a LocalCluster very much the same as your code does.  (Incidentally, your code has  `memory_target` as a LocalCluster param and the correct name is `memory_limit`, at least when I tried it on my workstation)

Here's my code:
```
    ngff_image = to_ngff_image(data,
                               dims= ("t", "c", "z", "y", "x"),
                               scale = {"c":1, "t":1, "x":im3.physical_pixel_sizes.X, "y":im3.physical_pixel_sizes.Y, "z":im3.physical_pixel_sizes.Z},
                               translation = None,
                               name = "image",
                               axes_units = None
    )
    multiscales = to_multiscales(
        ngff_image,
        method=Methods.DASK_IMAGE_NEAREST,
        progress=rich_dask_progress,
        chunks=zarr_chunk_dims[0],
        cache= (ngff_image.data.nbytes > config.memory_target),
        scale_factors=[{"x":2, "y":2, "z":1, "t":1, "c":1}]
    )
    output_store = DirectoryStore(f"s3://{output_bucket}/{output_filename}.zarr", dimension_separator="/")
    to_ngff_zarr(
        output_store,
        multiscales,
        progress=rich_dask_progress
    )
```
My input data is a stacked dask delayed array cobbled together from several images.
The output chunks are intended to be of shape (t:1, c:1, z:1, ysize, xsize) which is one slice per chunk.

**Comment by thewtex** [comment]
Thanks for the details, looks good.

A couple things to try,

- a different `method`
- not using LocalCluster
- specify `chunks` in `to_multiscales` to your desired chunking, i.e. t, c, z are 1

If it is a label image, #61 will (todo) be adding a `ITKWASM_LABEL_IMAGE` `method` that will have fewer artifacts and possibly better performance.

**Comment by thewtex** [comment]
> Incidentally, your code has memory_target as a LocalCluster param and the correct name is memory_limit, at least when I tried it on my workstation)

I have a hard time finding this, at least with `git grep memory_limit` -- is there a file and line number?

**Comment by GFleishman** [comment]
My two cents - I don't think this new discussion is related to the nominal issue of this thread (yet). @toloudis based on the RuntimeError you're getting it looks to me like dask thinks you want the reconstructed ome-ngff-zarr object returned in memory, instead of written to disk. Overall your array is bigger than the one I wrote this issue about - so even if you get past your RuntimeError you will still need to wait a long time for dask to parse the task graph before any workers actually start creating the new format data (though it also depends on the output chunk size, not just the input data size), but if you wait long enough it will work. I suggest you continue working with a LocalCluster, since that means you probably own the machine which is doing the work. That way you're not paying for cloud compute or cluster chargebacks while the dask scheduler does its long think.

I haven't used s3, but just wondering if your local machine is trying to gather all of the output data in memory before it's then written to the DirectoryStore?

**Comment by toloudis** [comment]
Regarding memory_target:
https://github.com/thewtex/ngff-zarr/blob/main/ngff_zarr/cli.py#L267
According to the error I got, `memory_target` is not a recognized kwarg.  But `memory_limit` is. (see https://distributed.dask.org/en/stable/api.html#distributed.LocalCluster)

And yes I agree my current issue is probably different than the original post here.  Happy to move it to discussion or a new issue.   

I am trying many different methods right now to make this work stably and fast.  In particular, when I use `ome-zarr-py` (whose dask rescaling I actually worked on) I do have the very long initial stage of building a huge dask graph before the computation starts and it starts writing.   I'm also experimenting with just writing my own code to loop over the large time series and serialize parts of the computation in my own way before handing things to dask. Currently not yet using all workers optimally but robustly working and flexible.

I also realized that zarr's DirectoryStore is not good for s3 writes -- FSStore is more appropriate. But haven't tried to re-run with that in place just yet.

**Comment by GFleishman** [comment]
Ahh - I knew I recognized your avatar - it must have been from various ome repos I've lurked through.

> ...experimenting with just writing my own code to loop over the large time series and serialize parts of the computation in my own way before handing things to dask

That's the approach that I've taken with a lot of my dask work. The pain point I guess is that if the distributed package makes great improvements to graph parsing and this kind of overhead issue goes away, then I'll have to rewrite things back into the intended dask work flow, i.e. use collections without having to think about how they work under the hood. That's definitely possible, dask/distributed are great tools with great developer teams. One approach is to jump in and lend a hand and try to make improvements from within. But my job is more about delivering processed results to stake holders, so taking long periods to work on code infrastructure is hard to justify. So I work around when I have to work around; and often the work arounds are pretty simple.

For example, `dask.array.map_overlap` scales really poorly - for every block is spawns maybe ~100 tasks. So if you have a really large 3D or 4D dataset and you want to do an overlapping computation, you'll wait hours for the scheduler to parse the graph. On the other hand, just sending the coordinates of the block plus its overlaps to each worker and having the worker read data out of the store directly is just one task per block - super easy - [and determining the block+overlap positions is simple.](https://github.com/JaneliaSciComp/bigstream/blob/6bd137df243dd31f3c4537d858b910d4e301d05e/bigstream/piecewise_align.py#L192)

**Comment by thewtex** [comment]
> According to the error I got, memory_target is not a recognized kwarg. But memory_limit is. (see https://distributed.dask.org/en/stable/api.html#distributed.LocalCluster)

Thanks! Patched in #63 

> specify chunks in to_multiscales to your desired chunking, i.e. t, c, z are 1 -- currently, they are all zarr_chunk_dims[0]

I am interested to know whether this addresses the behavior?

**Comment by dpshepherd** [comment]
I think we are running into the same problem. 

We are trying to create an ome-zarr that has 17 channels. Each channel is saved to disk as an independent zarr file. The shape of each independent zarr is `(69, 24989, 20210)`. 

The following is the code we are trying:
```python 
    print("\nLoading full-scale resolutions...")
    im_data = []
    for ch_idx in tqdm(range(len(channel_ids)),desc="channel"):
        full_res_path = root_path / Path(r"full_resolution") / Path("ch"+str(ch_idx).zfill(3)+".zarr")
        im_data.append(da.from_zarr(full_res_path))
        
    im_data = da.stack(im_data,axis=0)
    
    image = nz.to_ngff_image(
        im_data,
        dims=['c','z','y', 'x'],
        scale={'z' : .32,'y' : .098,'x': .098},
        translation={'z' : 1.0,'y' : 1.0,'x': 1.0}
    )
    
    print("\nCreating multiscale mapping...")
    with dask.diagnostics.ProgressBar():
        multiscales = nz.to_multiscales(
            image,
            scale_factors=[2,4,8,16,32],
            chunks={'c': 1, 'z': 1, 'y': 256, 'x': 256},
        )
```

With this strategy, we get the following output,

```bash
Converting full dataset fusion to multiscale ome.zarr...

Loading full-scale resolutions...
channel: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:00<00:00, 194.03it/s]

Creating multiscale mapping...
[########################################] | 100% Completed | 100.54 ms
[########################################] | 100% Completed | 100.33 ms
```

And the computer sits for hours. We left it alone for ~50 hours and it did not move beyond this step. It is a Ubuntu 22.04 LTS machine with 12-core AMD processor and 128 gb of RAM. Plenty of hard disk space.

Any suggestions for how to speed things up?

**Comment by thewtex** [comment]
@dpshepherd one approach may be to implement an optional [cubed](https://github.com/cubed-dev/cubed) backend in ngff-zarr to used instead of dask.

**Comment by ahnsws** [comment]
Hello, we may be running into similar issues as well. We have a `(37, 30000, 30000)` image that we are generating pyramidal images for, and ngff_zarr takes too long to finish.

Here are the results of various pyramid generation libraries:

```
memory usage is peak memory usage value as profiled by memray
execution time is slightly slower due to profiling by memray
ngff_zarr, cache=False,    ITKWASM,    2x3000x3000:    6s,   235MB
ngff_zarr, cache=False,    ITKWASM,   15x3000x3000:   29s,   448MB
ngff_zarr, cache=False,    ITKWASM,   30x3000x3000:   97s,   585MB

ngff_zarr, cache=False,    ITKWASM,  2x15000x15000:   26s,   731MB
ngff_zarr, cache=False,    ITKWASM, 15x15000x15000:  307s,  3395MB
ngff_zarr, cache=False,    ITKWASM, 30x15000x15000: 1000s,  6354MB

ngff_zarr, cache=False, DASK_IMAGE,    2x3000x3000:    8s,   262MB
ngff_zarr, cache=False, DASK_IMAGE,   15x3000x3000:   43s,   490MB
ngff_zarr, cache=False, DASK_IMAGE,   30x3000x3000:   91s,   796MB

ngff_zarr, cache=False, DASK_IMAGE,  2x15000x15000:  157s,  1257MB
ngff_zarr, cache=False, DASK_IMAGE, 15x15000x15000: 1048s,  8223MB
ngff_zarr, cache=False, DASK_IMAGE, 30x15000x15000: 2026s, 16248MB

xarray_multiscale, windowed_mean,      2x3000x3000:    2s,   235MB
xarray_multiscale, windowed_mean,     15x3000x3000:    4s,   349MB
xarray_multiscale, windowed_mean,     30x3000x3000:    8s,   347MB

xarray_multiscale, windowed_mean,    2x15000x15000:    9s,   414MB
xarray_multiscale, windowed_mean,   15x15000x15000:   51s,   420MB
xarray_multiscale, windowed_mean,   30x15000x15000:   95s,   451MB

ome-zarr, gaussian,                    2x3000x3000:    1s,   360MB
ome-zarr, gaussian,                   15x3000x3000:    3s,   524MB
ome-zarr, gaussian,                   30x3000x3000:    6s,   592MB

ome-zarr, gaussian,                  2x15000x15000:    8s,  1142MB
ome-zarr, gaussian,                 15x15000x15000:   34s,  1103MB
ome-zarr, gaussian,                 30x15000x15000:   59s,  1124MB
```

For the test dataset above, thankfully they all finished, but for our actual microscope images, some runs ran out of memory with ngff_zarr (the error was thrown by dask somewhere); this led me to investigate the issue further.

The code is here; it's a bit messy, forgive me, but it's the one I used to generate the results. If there are any other options I should try, please let me know, thanks.

```python

import glob
import time
from pathlib import Path

import dask.array as da
import ngff_zarr as nz
import numpy as np
import skimage.data
import xarray as xr
import zarr
import zarr.storage
from dask.diagnostics import ProgressBar
from natsort import natsorted
from ome_zarr.scale import Scaler
from ome_zarr.writer import write_image
from xarray_multiscale import windowed_mean, multiscale
from zarr.codecs import BloscCodec, BytesCodec


def create_synthetic_single_channel_zarrs():
    dst = Path(rf"./images")
    astronaut = skimage.data.astronaut()
    assert astronaut.ndim == 3  # YXC
    assert astronaut.shape[-1] == 3

    # WARNING: default values will create a large zarr locally
    tile_size = 2048

    for channel in range(30):
        astronaut_ch = np.tile(astronaut[..., channel % 3], (30, 30))
        out = zarr.open(
            str(dst / f"channel{channel}.zarr"),
            fill_value=0,
            shape=astronaut_ch.shape,
            dtype=astronaut.dtype,
            chunks=(tile_size, tile_size),
            codecs=[BytesCodec(), BloscCodec()],
            mode="w",
        )
        out[:, :] = astronaut_ch


def get_base_array() -> da.Array:
    channel_paths = glob.glob(rf"./images/channel*.zarr")
    channel_paths = natsorted(channel_paths)
    channel_paths = [Path(p) for p in channel_paths]

    # collect individual channel arrays
    print("Channel arrays:")
    arrays = []
    for i, path in enumerate(channel_paths):
        dask_array = da.from_zarr(zarr.open(str(path), mode="r"))
        arrays.append(dask_array)
        print(f"  {path.name}: {dask_array.shape}")
    base_array = da.stack(arrays, axis=0)
    assert base_array.ndim == 3  # CYX

    # base_array = base_array[:2, :3000, :3000]
    print(base_array.shape)
    return base_array


def _compute_scale_factors(
    image_shape: tuple, spatial_dims: list[str], top_level_size: int, downscale: int
) -> list[int]:
    # Get indices of spatial dimensions (not 'c' or 't')
    inds = [i for i, dim in enumerate(spatial_dims) if dim in ["y", "x", "z"]]
    sizes = [image_shape[i] for i in inds]
    max_size = max(sizes)

    # Compute how many times we need to downsample by 2
    scales = []
    current_scale = downscale
    while max_size // current_scale >= top_level_size:
        scales.append(current_scale)
        current_scale *= downscale

    # Always include one more level beyond the tile size threshold
    if scales:
        scales.append(current_scale)
    else:
        # no downsampling needed
        # eh do a more meaningful check here
        pass

    return scales


def generate_pyramid_ngff_itkwasm():
    base_array = get_base_array()

    # pyramid parameters
    dst = Path(rf"./pyramids/pyramid_ngff_itkwasm.zarr")
    downscale = 2
    tile_size = 2048
    max_top_level_size = 256

    # convert to NGFF image
    dims = ["c", "y", "x"]
    image = nz.to_ngff_image(
        base_array,
        dims=dims,
        scale={"c": 1.0, "y": 1.0, "x": 1.0},
        translation={"c": 0.0, "y": 0.0, "x": 0.0},
    )

    # compute scale factors
    scale_factors = _compute_scale_factors(
        image.data.shape, dims, max_top_level_size, downscale=downscale
    )
    print("scale factors:", scale_factors)

    # generate ngff-zarr multiscales
    print("generating multiscales...")  # stuck here when there are many channels
    with ProgressBar():
        multiscales = nz.to_multiscales(
            data=image,
            scale_factors=scale_factors,
            chunks={"c": 1, "y": tile_size, "x": tile_size},
            method=nz.Methods.ITKWASM_GAUSSIAN,
            # method=nz.Methods.DASK_IMAGE_GAUSSIAN,
            cache=False,
        )
    print(f"pyramid levels to be generated: {len(multiscales.images)}")
    for i, img in enumerate(multiscales.images):
        print(f"  level {i} (1:{2**i:>4}): {img.data.shape}")

    # now actually generate pyramid
    print("Writing pyramid...")
    with ProgressBar():
        nz.to_ngff_zarr(
            store=dst,
            multiscales=multiscales,
            overwrite=True,
        )
    print("pyramid written")


def generate_pyramid_xarray_multiscale():
    base_array = get_base_array()

    # pyramid parameters
    dst = Path(rf"./pyramids/pyramid_xarray_multiscales.zarr")
    downscale = 2
    tile_size = 2048
    max_top_level_size = 256

    print("Creating xarray multiscale")
    with ProgressBar():
        levels: list[xr.DataArray] = multiscale(
            base_array,
            reduction=windowed_mean,
            scale_factors=[1, downscale, downscale],
            chained=False,  # recompute from base layer
            preserve_dtype=True,
            chunks=[1, tile_size, tile_size],
        )

    # remove levels that are too small; not sure what option to pass to multiscale to
    # restrict this
    big_levels = []
    for level in levels:
        if max(level.shape) >= max_top_level_size:
            big_levels.append(level)

    print("Creating ngff images")
    dims = ["c", "y", "x"]
    ngff_images = []
    level: xr.DataArray
    for i, level in enumerate(big_levels):
        scale_factor = downscale**i
        ngff_images.append(
            nz.to_ngff_image(
                level,
                dims=dims,
                scale={"c": 1.0, "y": scale_factor, "x": scale_factor},
                translation={"c": 0, "y": 0, "x": 0},
                name="image",
            )
        )

    print("Writing pyramid levels")
    image: nz.NgffImage
    for index, image in enumerate(ngff_images):
        level_path = dst / f"scale{index}" / "image"
        with ProgressBar():
            image.data.to_zarr(level_path, mode="w", compute=True)


def generate_pyramid_ome_zarr():
    base_array = get_base_array()

    # pyramid parameters
    dst = Path(rf"./pyramids/pyramid_ome_zarr.zarr")
    downscale = 2
    tile_size = 2048
    max_top_level_size = 256

    # convert to NGFF image
    dims = ["c", "y", "x"]
    image = nz.to_ngff_image(
        base_array,
        dims=dims,
        scale={"c": 1.0, "y": 1.0, "x": 1.0},
        translation={"c": 0.0, "y": 0.0, "x": 0.0},
    )
    scale_factors = _compute_scale_factors(
        image.data.shape, dims, max_top_level_size, downscale=downscale
    )
    print("scale factors:", scale_factors)

    max_layers = len(scale_factors) + 1
    print("max layers", max_layers)
    # this is all we need the NGFF image for; argh refactor this

    root = zarr.open(str(dst), mode="w")

    write_image(
        image=base_array,
        group=root,
        axes="cyx",
        scaler=Scaler(
            method="gaussian",  # 'gaussian', 'local_mean', 'nearest'
            max_layer=max_layers,  # Number of pyramid levels
            downscale=downscale,  # 2x downsampling per level
        ),
        chunks=(1, tile_size, tile_size),
        compute=True,  # Compute immediately
    )


def run():
    t0 = time.perf_counter()
    generate_pyramid_ngff_itkwasm()
    # generate_pyramid_xarray_multiscale()
    # generate_pyramid_ome_zarr()
    t1 = time.perf_counter()
    # print(f"xarray-multiscale: {t1-t0:,.3f}s")
    print(f"ome-zarr: {t1-t0:,.3f}s")
    # print(f"ngff_zarr+dask_image: {t1-t0:,.3f}s")


if __name__ == "__main__":
    run()


```

**Comment by ahnsws** [comment]
I didn't want to use cache=True because I wanted to avoid writing potentially large amounts of data to my boot drive since it's not so large (my secondary drive is much larger). Regardless, running with cache=True doesn't really change the timings/memory usage much:

```
memory usage is peak memory usage value as profiled by memray
execution time is slightly slower due to profiling by memray
ngff_zarr, cache=False,    ITKWASM,    2x3000x3000:    6s,   235MB
ngff_zarr, cache=False,    ITKWASM,   15x3000x3000:   29s,   448MB
ngff_zarr, cache=False,    ITKWASM,   30x3000x3000:   97s,   585MB

ngff_zarr, cache=False,    ITKWASM,  2x15000x15000:   26s,   731MB
ngff_zarr, cache=False,    ITKWASM, 15x15000x15000:  307s,  3395MB
ngff_zarr, cache=False,    ITKWASM, 30x15000x15000: 1000s,  6354MB

ngff_zarr, cache=False, DASK_IMAGE,    2x3000x3000:    8s,   262MB
ngff_zarr, cache=False, DASK_IMAGE,   15x3000x3000:   43s,   490MB
ngff_zarr, cache=False, DASK_IMAGE,   30x3000x3000:   91s,   796MB

ngff_zarr, cache=False, DASK_IMAGE,  2x15000x15000:  157s,  1257MB
ngff_zarr, cache=False, DASK_IMAGE, 15x15000x15000: 1048s,  8223MB
ngff_zarr, cache=False, DASK_IMAGE, 30x15000x15000: 2026s, 16248MB

ngff_zarr, cache=True,     ITKWASM,    2x3000x3000:    5s,   248MB
ngff_zarr, cache=True,     ITKWASM,   15x3000x3000:   32s,   398MB
ngff_zarr, cache=True,     ITKWASM,   30x3000x3000:   88s,   531MB

ngff_zarr, cache=True,     ITKWASM,  2x15000x15000:   32s,   715MB
ngff_zarr, cache=True,     ITKWASM, 15x15000x15000:  327s,  3455MB
ngff_zarr, cache=True,     ITKWASM, 30x15000x15000: 1259s,  5837MB (crashed once)

ngff_zarr, cache=True,  DASK_IMAGE,    2x3000x3000:    7s,   271MB
ngff_zarr, cache=True,  DASK_IMAGE,   15x3000x3000:   49s,   490MB
ngff_zarr, cache=True,  DASK_IMAGE,   30x3000x3000:   92s,   797MB

ngff_zarr, cache=True,  DASK_IMAGE,  2x15000x15000:  148s,  1258MB
ngff_zarr, cache=True,  DASK_IMAGE, 15x15000x15000: 1027s,  8228MB
ngff_zarr, cache=True,  DASK_IMAGE, 30x15000x15000: 2067s, 16266MB

xarray_multiscale, windowed_mean,      2x3000x3000:    2s,   235MB
xarray_multiscale, windowed_mean,     15x3000x3000:    4s,   349MB
xarray_multiscale, windowed_mean,     30x3000x3000:    8s,   347MB

xarray_multiscale, windowed_mean,    2x15000x15000:    9s,   414MB
xarray_multiscale, windowed_mean,   15x15000x15000:   51s,   420MB
xarray_multiscale, windowed_mean,   30x15000x15000:   95s,   451MB

ome-zarr, gaussian,                    2x3000x3000:    1s,   360MB
ome-zarr, gaussian,                   15x3000x3000:    3s,   524MB
ome-zarr, gaussian,                   30x3000x3000:    6s,   592MB

ome-zarr, gaussian,                  2x15000x15000:    8s,  1142MB
ome-zarr, gaussian,                 15x15000x15000:   34s,  1103MB
ome-zarr, gaussian,                 30x15000x15000:   59s,  1124MB
```

Yes, it looks like xarray-multiscale and ome-zarr use different algorithms; what I use now that works well for us is ome-zarr but with a modified Scaler that blurs (using map_overlap) and then decimates, which runs a tiny bit longer than the timings for ome-zarr gaussian.

**Comment by ahnsws** [comment]
Ah yes so I'm using uv as my python environment manager; I installed memray and I ran the script using `uv run memray run script.py`, and once that finished, I ran `uv run memray stats memray-script.py.{numbers}.bin`, which printed a bunch of stats; I just wrote down the peak memory usage since for our actual microscopy images (37x30000x30000), we kept running into memory issues with ngff_zarr.

**Comment by jo-mueller** [comment]
Thanks! The dask dashboard really is a great introspection tool. 

>  it looks like the condition should be `index > 1?`

Tbh, I think I we could probably remove the index-related condition alltogether. For scale factors [2, 4, 8, 16], the function is first invoked for scale factor 2. At this time, the first level has already been written and *could* directly be loaded for the computation of the next scale level. 

I think especially in cases when the `NgffImage` data is the result of a delayed workflow, writing any level from the delayed dask graph would still drag along the weight of the entire previous workflow. Just writing the root resolution from the delayed dask graph and then creating the successive levels from the previously written data seems like a reasonable approach to me.

**Comment by jo-mueller** [comment]
I did some more tests (WSI image of dimensions `4 x 18k x 15k`, `"cyx"`) and despite the improvements in #280 I get an excessively large dask graph, i.e.:

- ~30k jobs for 0th-layer
- 330k for 1st layer
- `<tilt>` amount of jobs for 2nd layer with dask graph of size 270MB which is insane.

So there is still something in the code which gives a huge hit to performance. I tried a bit with a simple own implementation of `da.map_blocks(some_downsampling_function)` and get away with way less jobs. Will probably require more investigation to close this 🤔 😕

**Comment by thewtex** [comment]
@jo-mueller yes, for analysis, archival, or high-quality visualization purposes, we want gaussian filtering with the requisite overlap. For quick-and-dirty visualization, we can get away with using the bin-shrink algorithm, with the caveat that it will generate artifacts, especially for larger images. And, we can leverage that overlap is not required for the bin-shrink case and use map_blocks as a fast path. Note that to use the fast path there should be a check that the shape of the chunk is mod 0 for the shrink factor in each direction.

**Comment by thewtex** [comment]
Thanks all for the discussion on this issue. We made significant improvements. 

ngff-zarr currently handles larger-than-memory image conversions fairly well with Dask.

We will will make more improvements in the future to scalability and performance with refactoring that does not rely on Dask.
